src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/KeyboardEventSender.cs

```diff
@@ -24,83 +25,101 @@ public static void Reset()
          */
         public static void SendKeysInOneEvent(int replaySegment, List<(Key, KeyState)> keyStates)
         {
+            #if ENABLE_LEGACY_INPUT_MANAGER
+            SendKeysInOneEventLegacy(keyStates);
+            #endif
+            
             var time = InputState.currentTime;
             var keyboard = Keyboard.current;
 
             var keys = new HashSet<Key>();
+            NativeArray<byte>? deltaStateEventArray = null;
             InputEventPtr? deltaStateEvent = null;
-
-            List<(Key, KeyState)> logList = new();
-
-            foreach (var valueTuple in keyStates)
+            try
             {
-                var key = valueTuple.Item1;
-                var upOrDown = valueTuple.Item2;
+                List<(Key, KeyState)> logList = new();
 
-                if (deltaStateEvent.HasValue)
+                foreach (var valueTuple in keyStates)
                 {
-                    if (key == Key.LeftShift || key == Key.RightShift)
-                    {
-                        // start a new event
-                        _isShiftDown = upOrDown == KeyState.Down;
-                        RGDebug.LogInfo($"({replaySegment}) Sending Multiple Key Event: [" + string.Join(", ", logList.Select(a => a.Item1 + ":" + a.Item2).ToArray()) + "]");
-                        logList.Clear();
-                        InputSystem.QueueEvent(deltaStateEvent.Value);
-                        deltaStateEvent = null;
-                    }
-                    else
+                    var key = valueTuple.Item1;
+                    var upOrDown = valueTuple.Item2;
+
+                    if (deltaStateEvent.HasValue)
                     {
-                        if (keys.Contains(key))
+                        if (key == Key.LeftShift || key == Key.RightShift)
                         {
+                            // start a new event
+                            _isShiftDown = upOrDown == KeyState.Down;
                             RGDebug.LogInfo($"({replaySegment}) Sending Multiple Key Event: [" + string.Join(", ", logList.Select(a => a.Item1 + ":" + a.Item2).ToArray()) + "]");
                             logList.Clear();
                             InputSystem.QueueEvent(deltaStateEvent.Value);
                             deltaStateEvent = null;
-                            keys.Clear();
+                            deltaStateEventArray.Value.Dispose();
+                            deltaStateEventArray = null;
+                        }
+                        else
+                        {
+                            if (keys.Contains(key))
```

Comment:
This code I wrote still allows 2 events in the same update for the same keys.  Based on what you reported with the input system and how it handles events, would it be better to make this whole method only accept a dictionary of key states ?

If we wanted to still support a method with the current signature (Which would be nice so I don't have to move all this buffer logic into application code for starsreach), then if feels like we would need to manage our own key buffer across frames and hold off on building/sending the follow up event until the next frame.

Reply:
I like the idea of converting it to accept a dictionary. 

If we want to have an API that allows for sending multiple individual key events within one frame, then I think we do need to do some buffering ourselves (similar to what I'm doing with keyStates in the action manager), since we can't send delta events for individual keys directly to the Input System due to the bitfield issue.

Though I don't think this necessarily means we need to hold off until the next frame - it just means that if another call to our API is made, then we should queue another keyboard event to the Input System that includes any buffered key changes + the new key change.